src/torchcodec/_core/CudaDeviceInterface.cpp

```diff
@@ -401,127 +325,39 @@ void CudaDeviceInterface::convertAVFrameToFrameOutput(
   // also need to check that the AVFrame is in AV_PIX_FMT_NV12 format (8 bits),
   // because this is what the NPP color conversion routines expect. This SHOULD
   // be enforced by our call to maybeConvertAVFrameToNV12OrRGB24() above.
-  // TODONVDEC P2 this can be hit from the beta interface, but there's no
-  // hw_frames_ctx in this case. We should try to understand how that affects
-  // this validation.
-  AVHWFramesContext* hwFramesCtx = nullptr;
-  if (avFrame->hw_frames_ctx != nullptr) {
-    hwFramesCtx =
-        reinterpret_cast<AVHWFramesContext*>(avFrame->hw_frames_ctx->data);
-    AVPixelFormat actualFormat = hwFramesCtx->sw_format;
-
-    TORCH_CHECK(
-        actualFormat == AV_PIX_FMT_NV12,
-        "The AVFrame is ",
-        (av_get_pix_fmt_name(actualFormat) ? av_get_pix_fmt_name(actualFormat)
-                                           : "unknown"),
-        ", but we expected AV_PIX_FMT_NV12. "
-        "That's unexpected, please report this to the TorchCodec repo.");
-  }
+  TORCH_CHECK(
+      avFrame->hw_frames_ctx != nullptr,
+      "The AVFrame does not have a hw_frames_ctx. This should never happen");
+  AVHWFramesContext* hwFramesCtx =
+      reinterpret_cast<AVHWFramesContext*>(avFrame->hw_frames_ctx->data);
+  TORCH_CHECK(
+      hwFramesCtx != nullptr,
+      "The AVFrame does not have a valid hw_frames_ctx. This should never happen");
 
-  torch::Tensor& dst = frameOutput.data;
-  if (preAllocatedOutputTensor.has_value()) {
-    dst = preAllocatedOutputTensor.value();
-  } else {
-    dst = allocateEmptyHWCTensor(frameDims, device_);
-  }
+  AVPixelFormat actualFormat = hwFramesCtx->sw_format;
+  TORCH_CHECK(
+      actualFormat == AV_PIX_FMT_NV12,
+      "The AVFrame is ",
+      (av_get_pix_fmt_name(actualFormat) ? av_get_pix_fmt_name(actualFormat)
+                                         : "unknown"),
+      ", but we expected AV_PIX_FMT_NV12. "
+      "That's unexpected, please report this to the TorchCodec repo.");
 
```

Comment:
Below: we would previously do the NVDEC / NPP stream synchronization here. Now, it's done a bit later inside of `convertNV12FrameToRGB`. So this portion just becomes about figuring out what the `NVDEC` stream is so we can pass it down to `convertNV12FrameToRGB`.